### DISCUSSION TITLE
Accessing the template block  when using Phlex::DeferredRender

### DISCUSSION THREAD

**Discussion by stex** [yippee-fun/phlex#571] [Help] [answered]
Hey!

When designing new components for our application, I switched from `ViewComponent` to `phlex` now and quite like it so far :) 

I have a `Modal` component which can react to various request formats and behave accordingly:

```ruby
  def template(&)
    if !turbo_stream? && !turbo_frame?
      plain_content(&)
    elsif turbo_frame?
      frame_content(&)
    elsif turbo_stream?
      stream_content(&)
    end
  end
```

However, I now wanted to extend this component to support an optional header element
as a `ViewComponent`-like `with_header`:

```haml
= render Modal.new do |c|
  - c.with_header("something")

  %p I am in the modal body
```

I'm using `Phlex::DeferredRender` to delay rendering the template until the header part is known, but that gets rid of the block. I could add a `#with_content` method, but that seems like overhead and isn't compatible to the original modal component any more.

Is there a way to access the output of the `Modal.new`-block while using `DeferredRender` so I can
just `yield` it in the various methods like I do without the `DeferredRender`?

Thanks a lot in advance!

**Comment by joeldrapper** [comment, marked as answer]
Hey @stex, we decided against passing the block to the template method when using deferred render because it could likely lead to confusing interfaces. Take this for example:

```ruby
render Card.new do |c|
  p { "Before" }
  c.title { "Hello" }
  p { "After" }
end
```

If the card was implemented in the "builder style" like this, you would get the expected output.

```ruby
class Card < Phlex::HTML
  def template(&)
    article(&)
  end

  def title(&)
    h1(class: "title", &)
  end
end
```

But if the card was implemented with deferred render, both paragraphs would appear _after_ the title.

```ruby
class Card < Phlex::HTML
  include Phlex::DeferredRender

  def template
    article do
      h1(class: "title", &@title)
      yield
    end
  end

  def title(&block)
    @title = block
  end
end
```

We thought it would be better to encourage people to ignore the output of the block itself and instead use specific methods for each content slot. In this case, we could implement a `content` or `with_content` method.

```ruby
render Card.new do |c|
  c.title { "Hello" }

  c.content do
    p { "Hello" } 
    p { "World" }
  end
end
```

**Reply by stex** [reply]
That's what I thought, but I hoped there was a way around it 🙈 
Thanks a lot for the fast answer!

**Reply by joeldrapper** [reply]
We could find a way around it (in fact, that would just be a case of deleting some code), but I’m concerned it would lead to confusing interfaces. What do you think?

**Reply by joeldrapper** [reply]
Another option you could consider: the header could be an argument instead of a method.

```ruby
class Modal < Phlex::HTML
  def initialize(title:)
    @title = title
  end
  
  def template(&)
    h1(class: "title") { @title } if @title
    
    if !turbo_stream? && !turbo_frame?
      plain_content(&)
    elsif turbo_frame?
      frame_content(&)
    elsif turbo_stream?
      stream_content(&)
    end
  end
end
```

Usage

```ruby
render Modal.new(title: "Whatever") do
  ...
end
```

A limitation with this approach is the title is a String, so can't contain additional markup. I’ve been thinking about how the phlex-markdown gem could offer a way to permit limited inline markdown in these strings. Bold and italics are good examples (`title: "Hello **world**"`).

You could make the title argument accept a proc, but then you would need to handle it differently in the component.

**Reply by stex** [reply]
It requires the one using the component to know how it behaves if the application uses both styles (deferred, direct). 

For something like `#with_title`, I think it would be clear that it goes into a fixed slot. 
I'd personally go for a naming convention in my apps, e.g. `with_´ will be put into a specific position,  everything else will be exactly where's it's placed.

Really hard to say. I'd love the freedom of having both, but I can see how it might confuse people.

**Reply by stex** [reply]
> [Title as an argument]

Unfortunately, the header is an own component that accepts a few attributes. I could pass down nested hashes (or a proc), but both don't seem very clean to me.

**Reply by joeldrapper** [reply]
Ah, that makes sense. Keep in mind you can pass an initialised component around and render it later. You can also pass a proc to render later.

With an initialised component:

```ruby
render Modal.new(
  title: Modal::Title.new { "Hello" }
) do |m|
  # ...
end
```

With a proc:

```ruby
render Modal.new(title: -> { "Hello" }) do
  # ...
end
```

You could use like this:

```ruby
def template
  render @title
  # ...
end
```

Or even this:

```ruby
def template
  render Title.new(&@title)
  # ...
end
```

**Reply by joeldrapper** [reply]
> For something like #with_title, I think it would be clear that it goes into a fixed slot.

That makes sense. I guess we could enable it, and it's up to you whether you `yield` or not.

@willcosgrove what do you think? I seem to remember the view component team regretted enabling a default slot for some of the reasons mentioned above.

@joelhawksley did you end up removing this in 3.0?

**Reply by willcosgrove** [reply]
I think I have changed my opinion on this. I also came from a ViewComponent background and so I used to be on the side of providing the captured block, because that was what I was accustomed to. But I do think that this can be subtly confusing, and it can so easily be made unambiguous by defining a method for your main content block of your component.

Now I view it as a helpful constraint, nudging the component author toward a better interface. But that's just my 2¢.

**Reply by joeldrapper** [reply]
I think it does feel a little strange when it's the only slot method. For example, this feels redundant when there's no title. It's not terrible, though, and I tend to agree.

```ruby
render Card do |c|
  c.with_content do
    p { "Content" }
  end
end
```

**Reply by willcosgrove** [reply]
That's a fair point.

> we could enable it, and it's up to you whether you `yield` or not.

Maybe it's best to keep Phlex unopinionated here, like it is in most every other scenario.

**Reply by stex** [reply]
> I think it does feel a little strange when it's the only slot method. For example, this feels redundant when there's no title. It's not terrible, though, and I tend to agree.
> 
> ```ruby
> render Card do |c|
>   c.with_content do
>     p { "Content" }
>   end
> end
> ```

That was the main reason I asked the question. In most cases, I'll only have the body/content and having to put a slot method around it seemed inelegant. 

> Maybe it's best to keep Phlex unopinionated here, like it is in most every other scenario.

As phlex.fun currently already contains an own page about `DeferredRender`, maybe an info box / a short paragraph about this behaviour would be enough to allow `yield`ing?

**Reply by stex** [reply]
Hey again :) 
Did you come to a conclusion regarding this topic?

I'm currently a bit stuck with a few components and before I work around not being able to `yield`, I wanted to ask whether it might be allowed in a future version of `phlex`.

---

Edit: I switched away from `DeferredRender` for now even though it's a bit less elegant
